### DISCUSSION TITLE
PIMCTL tool not working though pimd daemon is running.

### DISCUSSION THREAD

**Discussion by SandipJunghare** [troglobit/pimd#203] [General]
Hi @troglobit ,

Please find below screenshot for my previous question. The screenshot are for other machine now.
Here i can start pimd daemon but not pimctl tool.

![pimd5](https://user-images.githubusercontent.com/88190551/135985567-eecab821-35ec-49d9-861c-746e9e3fe8f7.png)
![pimd4](https://user-images.githubusercontent.com/88190551/135985573-8d5b9c7a-5d88-4c38-8d45-6996cc8ce12f.png)
![pimd3](https://user-images.githubusercontent.com/88190551/135985580-53553c6a-fe84-4b58-b05e-75643e0b5109.png)

Do i need any changes in confing file or in any other file??

Thanks,
Sandip

**Comment by troglobit** [comment]
pimd needs to be able to read and write certain files to be able to communicate with the system.  The paths to these files are set by the configure script.  The defaults paths used by configure is in the /usr/local prefix, due to GNU tradition.  The paths need to match what your system uses.

You need to tell us how you've run configure.  E.g., the README recommends the following, which works on most modern Linux distributions:

    ./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var

When installed, pimd will look in /etc for pimd.conf, add store state files, like the IPC socket used by pimctl in /var ...

**Reply by SandipJunghare** [reply]
Hi,
i used below commands for installation and building available in Building from GIT section.
![image](https://user-images.githubusercontent.com/88190551/135989581-dc7c234d-9806-4aee-a225-ab3facf83efd.png)

git clone https://github.com/troglobit/pimd.git
cd pimd/
./autogen.sh
./configure && make








Also do u mean to say i need to run below commands?,

./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var
make
make DESTDIR=/tmp/pimd-2.3.2-1 install-strip


Thanks.

**Reply by troglobit** [reply]
I see. Yes, it depends on your system what directories you want to use for installing 3rd party programs.  If you want to use /usr/local (the default), which is used in the screenshot you reference, you also need to have the proper directory structure set up in /usr/local.  Which most standard Linux distros don't have populated. So I recommend using `./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var` instead.

Is this your first time with configure scripts and Linux?  This is sort of common knowledge and well described in the GNU documentation and other sources for autotools based projects.  

Why do you want to set the DESTDIR variable to a versioned directory with an older version when you're building the latest master?

**Reply by SandipJunghare** [reply]
yes, this is my first time with configure scripts and linux.

For DESTDIR, i was not aware of its purpose and use.

With ./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var     pimctl still not working.

**Reply by SandipJunghare** [reply]
Also for,
pimd: 16:11:31.476 Failed binding IPC socket, client disabled: No such file or directory

where i can find "pimd.sock"  file so that i can put it in correct folder that is /var.
or it generates with correct build process options??


Thanks.

**Reply by troglobit** [reply]
I see, well I cannot teach you all the basics required for building, installing, and using Linux programs. You'll have to look elsewhere for that. I suggest you read up  on configure scripts in general, the GNU autoconf/automake variants, and how they work.  You need to understand all the switches to the configure script and environment variables like DESTDIR to be able to install and use pimd (and other such projects).

The pimd.sock file used by pimd and pimctl is by default installed to the (localstatedir)/run/pimd.sock, so it's directly set up by the configure script.

**Reply by SandipJunghare** [reply]
Hi @troglobit ,

I can now able to build the repository and got going with pimd/pimctl binaries as well. It perfectly lists the multicast group addresses.
I attached the  sample pic of pim routing table.
Currently, i am more looking into public API's of your repository to get the same details which are presented by the pimctl tool.
I believe ipc.c is file for that.
 
More public API's info from you is appreciable.

Thanks,
Sandip.


![clip1](https://user-images.githubusercontent.com/88190551/136545166-dfce592d-f0c0-4401-833b-79df86981b86.png)

**Reply by troglobit** [reply]
The API between pimd and pimctl is private.  I do not support or condone the use of that for other applications since I need to be able to change that interface freely.
